File: src/main/java/com/github/morningwn/t0501_t1000/t0901_t1000/t0931_t0940/T0933.java
package com.github.morningwn.t0501_t1000.t0901_t1000.t0931_t0940;

import com.github.morningwn.prefab.Out;
import org.junit.Test;

/**
 * <a href="https://leetcode-cn.com/problems/number-of-recent-calls/">933. 最近的请求次数</a>
 *
 * @author moringwn
 * @create in 2022/5/6 16:11
 */
public class T0933 {

    @Test
    public void test() {
        RecentCounter recentCounter = new RecentCounter();

        Out.println(recentCounter.ping(1));
        Out.println(recentCounter.ping(100));
        Out.println(recentCounter.ping(3001));
        Out.println(recentCounter.ping(3002));

    }

    class RecentCounter {
        private final int[] queue;
        private int startIndex = 0;
        private int size = 0;

        public RecentCounter() {
            queue = new int[10001];
        }

        public int ping(int t) {
            queue[size++] = t;

            int start = t - 3000;
            while (startIndex <= size && queue[startIndex] < start) {
                startIndex++;
            }
            return size - startIndex;
        }
    }
}

File: src/main/scala/cn/morningwn/t0501_t1000/t0901_t1000/T0933.scala
package cn.morningwn.t0501_t1000.t0901_t1000

import org.junit.Test

/**
 * <a href="https://leetcode-cn.com/problems/number-of-recent-calls/">933. 最近的请求次数</a>
 *
 * @author moringwn
 * @create in 2022/5/6 16:33
 */
class T0933 {

    @Test
    def test(): Unit = {
        var obj = new RecentCounter()
        println(obj.ping(1))
        println(obj.ping(100))
        println(obj.ping(3001))
        println(obj.ping(3002))
    }

    class RecentCounter() {
        private val queue = new Array[Int](10001);
        private var startIndex = 0;
        private var size = 0;

        def ping(t: Int): Int = {
            val start = t - 3000;
            queue(size) = t
            size += 1
            while (startIndex <= size && queue(startIndex) < start) {
                startIndex += 1
            }

            size - startIndex
        }

    }

}
